docs/source/guide/events.rst

```diff
@@ -0,0 +1,445 @@
+    client2.meta.events.register('event', my_second_function)
+
+    events_called = []
+    client1.meta.events.emit('event', my_list=events_called)
+    print(events_called)
+
+    events_called = []
+    client2.meta.events.emit('event', my_list=events_called)
+    print(events_called)
+
+Here is the output for the code sample::
+
+    ['Called my first function']
+    ['Called my second function']
+
+As shown in the code sample, when a client emits an event with its event
+emitter, it only called handlers registered to that event and event emitter.
+
+When clients are created, their event emitters are created by copying the
+event emitter from the session that created the client. Note the session's
+event emitter is accessed by its ``events`` attribute. So client event
+emitters will actually inherit the session's registered handlers when the
+client is instantiated. However after instantiation of the client, the
+session's and client's event emitters are seperate. Here is a code sample
+that illustrates this::
+
+    import boto3
+    from boto3.session import Session
+
+
+    def my_session_function(my_list, **kwargs):
+        my_list.append('Called my session function')
+
+    def my_other_session_function(my_list, **kwargs):
+        my_list.append('Called my other session function')
+
+    def my_client_function(my_list, **kwargs):
+        my_list.append('Called my client function')
+
+    session = Session()
+    session.events.register('event', my_session_function)
+
+    client = session.client('s3')
+    # Register a new handler to the client.
+    client.meta.events.register('event', my_client_function)
+
+    # Register a new handler after instantiation of the client.
+    session.events.register('event', my_other_session_function)
+
+    events_called = []
+    client.meta.events.emit('event', my_list=events_called)
+    print(events_called)
+
+    events_called = []
+    session.events.emit('event', my_list=events_called)
+    print(events_called)
+
+
+Here is the output for this code sample::
+
+  ['Called my session function', 'Called my client function']
+  ['Called my session function', 'Called my other session function']
+
+
+As expected, the client inherited the ``my_session_function`` registration
+because it was registered to the session prior to instantiation of the
+client. However, upon instantiation of the client, the event emitters of both
+the session and the client are isolated. The ``my_client_function`` was only
+registered to the client's event emitter so it was only called when the
+client emitted ``'event'``. Likewise, the ``my_other_session_function`` was
+registered after the creation of the client so it was only called when the
+session emitted ``'event'``.
+
+
+Boto3 Specific Events
+---------------------
+
+In the examples from the previous sections, they relied on creating and
+emitting your own event, which is a completely valid thing to do. However
```

Comment:
Again, I think we should deemphasize this.  This seems like an advanced use case.